crates/build/src/solidity_metadata.rs

```diff
@@ -0,0 +1,292 @@
+// Copyright (C) ink! contributors.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+mod abi;
+mod natspec;
+
+use std::{
+    collections::HashMap,
+    fs::{
+        self,
+        File,
+    },
+    path::{
+        Path,
+        PathBuf,
+    },
+};
+
+use alloy_json_abi::JsonAbi;
+use anyhow::{
+    Context,
+    Result,
+};
+use cargo_metadata::TargetKind;
+use contract_metadata::{
+    CodeHash,
+    Contract,
+    Source,
+};
+use ink_metadata::InkProject;
+use serde::{
+    de,
+    ser::SerializeMap,
+    Deserialize,
+    Serialize,
+};
+
+use self::natspec::{
+    DevDoc,
+    UserDoc,
+};
+use crate::{
+    code_hash,
+    CrateMetadata,
+};
+
+// Re-exports abi utilities.
+pub use self::abi::{
+    abi_path,
+    generate_abi,
+    write_abi,
+};
+
+/// Solidity compatible smart contract metadata.
+///
+/// Ref: <https://docs.soliditylang.org/en/latest/metadata.html>
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct SolidityContractMetadata {
+    /// Details about the compiler.
+    compiler: Compiler,
+    /// Source code language
+    language: String,
+    /// Generated information about the contract.
+    pub output: Output,
+    /// Compiler settings.
+    /// Required by the spec, but very Solidity/EVM specific.
+    // TODO: (@davidsemakula) include ink! compiler settings instead?
```

Comment:
The `cargo-contract` version that was used to compile would be fitting here.

Reply:
Will do. 
What do you think of using the key `settings.ink` add including the equivalent of [`source.build_info` from ink!'s metadata](https://use.ink/basics/metadata#source) for this? 
We're using the `ink` as a form of crude namespacing since this is technically not part of the Solidity metadata spec.

Reply:
This is done, but also added the code hash (for the binary) and the docker image (if any). See implementation.